improve shutdown perf with lots of LOBs

grandinj opened the pull request:
by changing the structure of the lobMap, and encoding the tableId in the lobId.
This lets us perform a very limited range scan when we need to do cleanup.

This is related to
  Very high DB opening/closing times (#3066)

grandinj committed:
improve shutdown perf with lots of LOBs

by changing the structure of the lobMap, and encoding the tableId in the lobId.
This lets us perform a very limited range scan when we need to do cleanup.

This is related to
  Very high DB opening/closing times (#3066)

grandinj committed:
fixes for the composite table/lob key

grandinj commented:
@andreitokar does this look reasonable?

grandinj committed:
fix Recover for lob data change

wburzyns commented:
I tried it by manually applying it to the current master, which is fine.

The change breaks the engine:
org.h2.jdbc.JdbcSQLNonTransientException: General error: "java.lang.IllegalArgumentException: The value may not be null [2.0.201/0]"; SQL statement: INSERT INTO MATDB_DATA VALUES(?, ?, ?) [50000-201]

wburzyns commented:
BTW Both loops that I mention in #3066 are O(n) with regard to the number of LOBs. This linear scaling makes LOBs in H2 virtually useless.

I'm not familiar with the internals of the MVStore, so pardon me if the following question is silly:
The O(n) loop in the constructor, that determines maxLobId - couldn't this value be cached somewhere and just read during opening?

grandinj committed:
fix bugs in shifting and in calling ceilingKey

(1) Need to cast tableId to long before shifting
(2) since our keys can be negative, need to use Long.MIN_VALUE

grandinj committed:
Revert "fix bugs in shifting and in calling ceilingKey"

This reverts commit c869180398bd3c2cb54ea7340942ab83439e4577.

grandinj committed:
fix bugs in shifting and in calling ceilingKey

(1) Need to cast tableId to long before shifting
(2) since our keys can be negative, need to use Long.MIN_VALUE

andreitokar commented on h2/src/main/org/h2/mvstore/db/LobStorageMap.java:
```diff
@@ -319,15 +313,13 @@ public void removeAllForTable(int tableId) {
-                    list.add(e.getKey());
-                }
+            Long startKey = Long.valueOf(tableId << 48);
+            Long endKey = Long.valueOf((tableId + 1) << 48);
```
Same problem with shifting int, probably should be:
```
            long startKey = (long)tableId << 48;
            long endKey = (long)(tableId + 1) << 48;
```

grandinj committed:
need to cast to long before shifting

grandinj committed:
add some range checks, just in case

someone tries something weird and creates too many lobs, or too many tables

grandinj committed:
fix scan for max lob unique id

grandinj committed:
remove gc of lobs at startup

I don't believe this is necessary any more since @andreitokar improved this class in commit ebc4cfc8cb1edd51ae5a0b17417e11d7bd89c9be

andreitokar reviewed (approved):
LGTM

grandinj commented:
@wburzyns can you try this now?

wburzyns commented:
All my unit tests are OK.

The 201+ DB I created a few weeks ago opened without any complaints in 36 sec (so no better than previously; is this expected?). Data retrieval was not possible due to java.lang.ClassCastException. Th DB closing took 1.5 sec (way better than previously). I was able to open the DB again.

I guess that the the on-disk format has changed. I'll try completely fresh import on weekend.

wburzyns commented:
I've looked at the new LobStorageMap's constructor. The O(n) loop is still present. Is there really no way to cache maxLobUniqueId?
